sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/QueryPlanConstraints.scala

```diff
@@ -78,6 +91,72 @@ trait ConstraintHelper {
     inferredConstraints -- constraints
   }
 
+  /**
+   * Infers an additional set of constraints from a given set of inequality constraints.
+   * For e.g., if an operator has constraints of the form (`a > b`, `b > 5`), this returns an
+   * additional constraint of the form `a > 5`.
+   */
+  def inferInequalityConstraints(constraints: Set[Expression]): Set[Expression] = {
+    val binaryComparisons = constraints.filter {
+      case _: GreaterThan => true
+      case _: GreaterThanOrEqual => true
+      case _: LessThan => true
+      case _: LessThanOrEqual => true
+      case _: EqualTo => true
+      case _ => false
+    }
+
+    val greaterThans = binaryComparisons.map {
+      case EqualTo(l, r) if l.foldable => EqualTo(r, l)
+      case LessThan(l, r) => GreaterThan(r, l)
+      case LessThanOrEqual(l, r) => GreaterThanOrEqual(r, l)
+      case other => other
+    }
+
+    val lessThans = binaryComparisons.map {
+      case EqualTo(l, r) if l.foldable => EqualTo(r, l)
+      case GreaterThan(l, r) => LessThan(r, l)
+      case GreaterThanOrEqual(l, r) => LessThanOrEqual(r, l)
+      case other => other
+    }
+
+    var inferredConstraints = Set.empty[Expression]
+    greaterThans.foreach {
+      case op @ BinaryComparison(source: Attribute, destination: Expression)
+        if destination.foldable =>
```

Comment:
I think that the foldability is not needed here. The new constraints do not have to only involve constants, but also any attribute.

Reply:
To avoid generating too many constraints. For example: `a > b > c > 1`. The expected inferred constraints are: `a > 1 and b > 1`. ` a > c` is useless.

Reply:
If a and c are in tihe same side of a join, then it can be pushed down.

Reply:
If a and c are in tihe same side of a join, then it can be pushed down.

Reply:
How to push down `a > c` if  both `a` and `c` are not foldable?

Reply:
I'm sorry, I used a wrong word. I meant pushed through the join into one of the sides.